main/ahrs.go

```diff
@@ -0,0 +1,192 @@
+		recipNorm = invSqrt(s0 * s0 + s1 * s1 + s2 * s2 + s3 * s3) // normalise step magnitude
+		s0 *= recipNorm
+		s1 *= recipNorm
+		s2 *= recipNorm
+		s3 *= recipNorm
+
+		// Apply feedback step
+		qDot1 -= beta * s0
+		qDot2 -= beta * s1
+		qDot3 -= beta * s2
+		qDot4 -= beta * s3
+	}
+    
+    // Integrate rate of change of quaternion to yield quaternion
+	q0 += qDot1 * (1.0 / sampleFreq)
+	q1 += qDot2 * (1.0 / sampleFreq)
+	q2 += qDot3 * (1.0 / sampleFreq)
+	q3 += qDot4 * (1.0 / sampleFreq)
+
+	// Normalise quaternion
+	recipNorm = invSqrt(q0 * q0 + q1 * q1 + q2 * q2 + q3 * q3)
+	q0 *= recipNorm
+	q1 *= recipNorm
+	q2 *= recipNorm
+	q3 *= recipNorm
+}
+
+func AHRSupdateIMU(gx, gy, gz, ax, ay, az float32)
+{
+    recipNorm float32
+	s0, s1, s2, s3 float32
+	qDot1, qDot2, qDot3, qDot4 float32
+	_2q0, _2q1, _2q2, _2q3, _4q0, _4q1, _4q2 ,_8q1, _8q2, q0q0, q1q1, q2q2, q3q3 float32
+
+	// Rate of change of quaternion from gyroscope
+	qDot1 = 0.5 * (-q1 * gx - q2 * gy - q3 * gz)
+	qDot2 = 0.5 * (q0 * gx + q2 * gz - q3 * gy)
+	qDot3 = 0.5 * (q0 * gy - q1 * gz + q3 * gx)
+	qDot4 = 0.5 * (q0 * gz + q1 * gy - q2 * gx)
+
+	// Compute feedback only if accelerometer measurement valid (avoids NaN in accelerometer normalisation)
+	if(!((ax == 0.0) && (ay == 0.0) && (az == 0.0))) {
+
+		// Normalise accelerometer measurement
+		recipNorm = invSqrt(ax * ax + ay * ay + az * az)
+		ax *= recipNorm
+		ay *= recipNorm
+		az *= recipNorm   
+
+		// Auxiliary variables to avoid repeated arithmetic
+		_2q0 = 2.0 * q0
+		_2q1 = 2.0 * q1
+		_2q2 = 2.0 * q2
+		_2q3 = 2.0 * q3
+		_4q0 = 4.0 * q0
+		_4q1 = 4.0 * q1
+		_4q2 = 4.0 * q2
+		_8q1 = 8.0 * q1
+		_8q2 = 8.0 * q2
+		q0q0 = q0 * q0
+		q1q1 = q1 * q1
+		q2q2 = q2 * q2
+		q3q3 = q3 * q3
+
+		// Gradient decent algorithm corrective step
+		s0 = _4q0 * q2q2 + _2q2 * ax + _4q0 * q1q1 - _2q1 * ay
+		s1 = _4q1 * q3q3 - _2q3 * ax + 4.0 * q0q0 * q1 - _2q0 * ay - _4q1 + _8q1 * q1q1 + _8q1 * q2q2 + _4q1 * az
+		s2 = 4.0 * q0q0 * q2 + _2q0 * ax + _4q2 * q3q3 - _2q3 * ay - _4q2 + _8q2 * q1q1 + _8q2 * q2q2 + _4q2 * az
+		s3 = 4.0 * q1q1 * q3 - _2q1 * ax + 4.0 * q2q2 * q3 - _2q2 * ay
+		recipNorm = invSqrt(s0 * s0 + s1 * s1 + s2 * s2 + s3 * s3) // normalise step magnitude
+		s0 *= recipNorm
+		s1 *= recipNorm
+		s2 *= recipNorm
+		s3 *= recipNorm
+
+		// Apply feedback step
+		qDot1 -= beta * s0
+		qDot2 -= beta * s1
+		qDot3 -= beta * s2
+		qDot4 -= beta * s3
+	}
+
+	// Integrate rate of change of quaternion to yield quaternion
+	q0 += qDot1 * (1.0 / sampleFreq)
+	q1 += qDot2 * (1.0 / sampleFreq)
+	q2 += qDot3 * (1.0 / sampleFreq)
+	q3 += qDot4 * (1.0 / sampleFreq)
+
+	// Normalise quaternion
+	recipNorm = invSqrt(q0 * q0 + q1 * q1 + q2 * q2 + q3 * q3)
+	q0 *= recipNorm
+	q1 *= recipNorm
+	q2 *= recipNorm
+	q3 *= recipNorm
+}
+

```

Comment:
Yes it is. See here:  https://en.m.wikipedia.org/wiki/Fast_inverse_square_root

I wasn't sure if go retained enough C that it would work. Alternatively, we could import the C code directly, which is what I originally tried, but couldn't get that to work. I'm sure I'm just over looking something since I'm new to go though...